save-preprocessor/src/main/kotlin/org/cqfn/save/preprocessor/controllers/DownloadProject.kt

```diff
@@ -0,0 +1,77 @@
+package org.cqfn.save.preprocessor.controllers
+
+import org.cqfn.save.preprocessor.Response
+import org.cqfn.save.repository.GitRepository
+
+import org.eclipse.jgit.api.Git
+import org.eclipse.jgit.api.errors.GitAPIException
+import org.eclipse.jgit.api.errors.InvalidRemoteException
+import org.eclipse.jgit.api.errors.TransportException
+import org.eclipse.jgit.transport.CredentialsProvider
+import org.eclipse.jgit.transport.UsernamePasswordCredentialsProvider
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.http.HttpStatus
+import org.springframework.http.ResponseEntity
+import org.springframework.web.bind.annotation.PostMapping
+import org.springframework.web.bind.annotation.RequestBody
+import org.springframework.web.bind.annotation.RestController
+import reactor.core.publisher.Mono
+import java.io.File
+
+/**
+ * A Spring controller for git project downloading
+ */
+@RestController
+class DownloadProject {
+    private val log = LoggerFactory.getLogger(DownloadProject::class.java)
+
+    @Value("\${save.repository}")
+    private lateinit var volumes: String
+
+    /**
+     * @param gitRepository - Dto of repo information to clone
+     * @return response entity with text
+     */
+    @Suppress("TooGenericExceptionCaught", "TOO_LONG_FUNCTION")
+    @PostMapping(value = ["/upload"])
+    fun upload(@RequestBody gitRepository: GitRepository): Response {
```

Comment:
We have decided that cloning should occur in background. It may take a lot of time, so this handler should return reposnse immediately (indicating that it has started execution) and then start cloning in background. It can be acieved like this:
```kotlin
return Mono.just( /* something with OK status */ )
    .subscribeOn(Schedulers.boundedElastic())
    .also {
        it.subscribe {
            // clone repo here
        }
    }
```